src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/PreferMethodReferenceCheck.java

```diff
@@ -0,0 +1,344 @@
+        TokenTypes.LITERAL_THROW,
+        TokenTypes.LITERAL_TRY,
+        TokenTypes.LITERAL_WHILE
+    );
+
+    @Override
+    public int[] getDefaultTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return new int[] {TokenTypes.LAMBDA};
+    }
+
+    @Override
+    public void visitToken(DetailAST ast) {
+        final List<String> args = getLambdaArgs(ast);
+        final DetailAST expr = ast.getLastChild();
+        final boolean hasViolation = checkExpr(expr, args);
+        if (hasViolation) {
+            log(ast, MSG_METHOD_REF);
+        }
+    }
+
+    /**
+     * Checks if this expression can be replaced with method reference.
+     *
+     * @param expr token to check
+     * @param args lambda arguments
+     * @return true if can be replaced, false otherwise
+     */
+    private static boolean checkExpr(DetailAST expr, List<String> args) {
+        final DetailAST child = expr.getFirstChild();
+        final boolean result;
+        switch (child.getType()) {
+            case TokenTypes.METHOD_CALL:
+                result = checkMethodCall(child, args);
+                break;
+            case TokenTypes.LITERAL_NEW:
+                result = checkObjectCreation(child, args);
+                break;
+            case TokenTypes.SLIST:
+            case TokenTypes.LITERAL_RETURN:
+                result = child.getFirstChild().getType() != TokenTypes.SEMI
+                    && checkExpr(child.getFirstChild(), args);
+                break;
+            default:
+                result = shouldCheck(expr)
+                    && checkExpr(expr.findFirstToken(TokenTypes.EXPR), args);
+                break;
+        }
+        return result;
+    }
+
+    /**
+     * Checks if given ast should be checked.
+     * Ast should have only one child expr, and have no other tokens from
+     * {@link #TOKEN_TYPES} list.
+     *
+     * @param ast token to check
+     * @return true if there is only one child expression, false otherwise
+     */
+    private static boolean shouldCheck(DetailAST ast) {
+        return ast.getChildCount(TokenTypes.EXPR) == 1
+            && ast.getChildCount(TokenTypes.SEMI) == 1
+            && !TokenUtil.findFirstTokenByPredicate(ast,
+                expr -> TOKEN_TYPES.contains(expr.getType())).isPresent();
+    }
+
+    /**
+     * Check that creation of the object can be replaced with constructor reference.
+     *
+     * @param ast token to check
+     * @param args lambda arguments
+     * @return true if token can be replaced, false otherwise
+     */
+    private static boolean checkObjectCreation(DetailAST ast, List<String> args) {
+        final DetailAST elist = ast.findFirstToken(TokenTypes.ELIST);
+        final DetailAST arrayDecl = ast.findFirstToken(TokenTypes.ARRAY_DECLARATOR);
```

Comment:
can we split this method into two methods?
 it look like check two separate concerns: object and array

Reply:
done